include/ck/tensor_operation/gpu/device/device_contraction_xdl_cshuffle.hpp

```diff
@@ -0,0 +1,802 @@
+          index_t CShuffleBlockTransferScalarPerVector_NPerBlock,
+          LoopScheduler LoopSched = make_default_loop_scheduler()>
+struct DeviceContraction_Xdl_CShuffle : public DeviceContraction<NumDimM,
+                                                                 NumDimN,
+                                                                 NumDimK,
+                                                                 AElementwiseOperation,
+                                                                 BElementwiseOperation,
+                                                                 CElementwiseOperation>
+{
+    using DeviceOp = DeviceContraction_Xdl_CShuffle;
+
+    static constexpr auto I0 = Number<0>{};
+    static constexpr auto I1 = Number<1>{};
+    static constexpr auto I2 = Number<2>{};
+
+    // assume A[M0, M1, M2, ..., K0, K1, K2...]
+    static auto MakeAGridDescriptor_AK0_M_AK1(const std::vector<index_t>& a_lengths_vec,
+                                              const std::vector<index_t>& a_strides_vec)
+    {
+        assert(a_lengths_vec.size() == NumDimM + NumDimK &&
+               a_strides_vec.size() == NumDimM + NumDimK);
+
+        const auto to_tuple = [&](auto& vec, auto Num) {
+            return generate_tuple([&](auto i) { return vec[i]; }, Num);
+        };
+
+        const auto a_lengths = to_tuple(a_lengths_vec, Number<NumDimM + NumDimK>{});
+        const auto a_strides = to_tuple(a_strides_vec, Number<NumDimM + NumDimK>{});
+
+        // dimension Ids for M0, M1, ...
+        constexpr auto mDimIds = typename arithmetic_sequence_gen<0, NumDimM, 1>::type{};
+
+        // dimension Ids for K0, K1, ...
+        constexpr auto kDimIds =
+            typename arithmetic_sequence_gen<NumDimM, NumDimM + NumDimK, 1>::type{};
+
+        // lengths for M0, M1, ...
+        const auto mLengths = get_container_subset(a_lengths, mDimIds);
+
+        // lengths for K0, K1, ...
+        const auto kLengths = get_container_subset(a_lengths, kDimIds);
+
+        // naive tensor A[M0, M1, M2, ..., K0, K1, K2...]
+        const auto a_grid_desc_ms_ks = make_naive_tensor_descriptor(a_lengths, a_strides);
+
+        // transformed tensor A[MRaw = M0 * M1 * M2 * ... , KRaw = K0 * K1 * K2 * ...]
+        const auto a_grid_desc_mraw_kraw = transform_tensor_descriptor(
+            a_grid_desc_ms_ks,
+            make_tuple(make_merge_transform(mLengths), make_merge_transform(kLengths)),
+            make_tuple(mDimIds, kDimIds),
+            make_tuple(Sequence<0>{}, Sequence<1>{}));
+
+        const auto MRaw = a_grid_desc_mraw_kraw.GetLength(I0);
+        const auto KRaw = a_grid_desc_mraw_kraw.GetLength(I1);
+
+        const auto M = math::integer_divide_ceil(MRaw, MPerBlock) * MPerBlock;
+        const auto K = math::integer_divide_ceil(KRaw, KPerBlock) * KPerBlock;
+
+        const auto MPad = M - MRaw;
+        const auto KPad = K - KRaw;
+
```

Comment:
change to mPad , kPad?

Reply:
I use ```mPad``` for index, and ```MPad``` for length.

I think we need to came up with a consistent naming style for dimension, index, and tensor descriptor. If you have a good idea, please write it down and let's discuss.